rclcpp_lifecycle/src/state.cpp

```diff
@@ -14,49 +14,99 @@
 
 #include "rclcpp_lifecycle/state.hpp"
 
-#include <lifecycle_msgs/msg/state.hpp>
+#include <string>
 
-#include <rcutils/allocator.h>
-#include <rcutils/strdup.h>
-#include <rcl_lifecycle/data_types.h>
+#include "lifecycle_msgs/msg/state.hpp"
 
-#include <string>
+#include "rcutils/allocator.h"
+#include "rcutils/strdup.h"
+
+#include "rcl_lifecycle/data_types.h"
 
 namespace rclcpp_lifecycle
 {
 
-State::State()
-: State(lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN, "unknown")
+void
+copy_from(State & lhs, const State & rhs)
+{
+  if (lhs.owns_rcl_state_handle_) {
+    auto state_handle = reinterpret_cast<rcl_lifecycle_state_t *>(
+      rhs.allocator_.allocate(sizeof(rcl_lifecycle_state_t), rhs.allocator_.state));
+    state_handle->id = rhs.state_handle_->id;
+    state_handle->label = rcutils_strndup(
+      rhs.state_handle_->label,
+      strlen(rhs.state_handle_->label),
+      rhs.allocator_);
+    lhs.state_handle_ = state_handle;
+  } else {
+    lhs.state_handle_ = rhs.state_handle_;
+  }
+}
+
+State::State(rcutils_allocator_t allocator)
+: State(lifecycle_msgs::msg::State::PRIMARY_STATE_UNKNOWN, "unknown", allocator)
 {}
 
-State::State(uint8_t id, const std::string & label)
-: owns_rcl_state_handle_(true)
+State::State(
+  uint8_t id,
+  const std::string & label,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_state_handle_(true)
 {
   if (label.empty()) {
     throw std::runtime_error("Lifecycle State cannot have an empty label.");
   }
 
-  auto state_handle = new rcl_lifecycle_state_t;
+  auto state_handle = reinterpret_cast<rcl_lifecycle_state_t *>(
+    allocator_.allocate(sizeof(rcl_lifecycle_state_t), allocator_.state));
   state_handle->id = id;
   state_handle->label =
-    rcutils_strndup(label.c_str(), label.size(), rcutils_get_default_allocator());
+    rcutils_strndup(label.c_str(), label.size(), allocator_);
 
   state_handle_ = state_handle;
 }
 
-State::State(const rcl_lifecycle_state_t * rcl_lifecycle_state_handle)
-: owns_rcl_state_handle_(false)
+State::State(
+  const rcl_lifecycle_state_t * rcl_lifecycle_state_handle,
+  rcutils_allocator_t allocator)
+: allocator_(allocator),
+  owns_rcl_state_handle_(false)
 {
   state_handle_ = rcl_lifecycle_state_handle;
 }
 
+State::State(const State & rhs)
+: allocator_(rhs.allocator_),
+  owns_rcl_state_handle_(rhs.owns_rcl_state_handle_)
```

Comment:
`state_handle_` is not being initialized here. Therefore the function called later can't safely determine if the handle was valid and might need to be deallocate before being overwritten.

Reply:
i don't understand this. The `state_handle_` is getting initialize in the `copy_from` function. What is being overwritten inside the copy constructor?

Reply:
The `copy_from` method is public and can be called for any instance. It must cleanup previously allocated memory as mentioned in the other comment. And that will require the member variables to be initialized here *before* calling the function.

Reply:
yeah, changed to a protected member function now.